### PR TITLE
Use grep in Docker regexes instead of bash

### LINKED ISSUE #77862: Optimize loop in elasticsearch-env
The loop where the environment variables are being 'filtered' contains a regex check performed by Bash. This can be really slow when dealing with a large amount of environment variables, adding minutes(!) to the start up time.

```bash
  while IFS='=' read -r envvar_key envvar_value
  do
    # Elasticsearch settings need to have at least two dot separated lowercase
    # words, e.g. `cluster.name`
    if [[ "$envvar_key" =~ ^[a-z0-9_]+\.[a-z0-9_]+ ]]; then
      if [[ ! -z $envvar_value ]]; then
        es_opt="-E${envvar_key}=${envvar_value}"
        es_arg_array+=("${es_opt}")
      fi
    fi
  done <<< "$(env)"
```

We currently patch this by replacing the `elasticsearch-env` file in the Docker image:
```
FROM elasticsearch:7.13.3
COPY elasticsearch-env /usr/share/elasticsearch/bin/
```

, where the loop has been modified as follows:
```
  while IFS='=' read -r envvar_key envvar_value
  do
    if [[ ! -z $envvar_value ]]; then
      echo "adding option ${envvar_key}=${envvar_value} to list of arguments"
      es_opt="-E${envvar_key}=${envvar_value}"
      es_arg_array+=("${es_opt}")
    fi
  # Elasticsearch settings need to have at least two dot separated lowercase words, e.g. `cluster.name`
  done <<< "$(env | egrep '^[a-z0-9_]+\.[a-z0-9_]+')"
```
This:
- takes the regex _out_ of the loop, and
- has the regex being performed by _egrep_ instead of _bash_ (=faster)

### PR DESCRIPTION
**Pull request by jeff-schram**
Closes #77862.

Optimize Docker startup times in environments with a lot of environment variables by using `grep` for regular expressions instead of `bash`.

### CODE REVIEW

**Comment by jeff-schram** [comment]
Not quite sure what to do with/about the Contributor Agreement. I signed it on the 8th and again (after seeing the _cla-checker-service_'s comment) today (22nd).

**Comment by mark-vieira** [comment]
@elasticmachine test this please

**Comment by mark-vieira** [comment]
@elasticmachine update branch

**Comment by mark-vieira** [comment]
@elasticmachine test this please

**Comment by pugnascotia** [comment]
@elasticmachine ok to test

**Comment by pugnascotia** [comment]
> @pugnascotia Hey Rory, was working on that as well. I was also wondering if the trailing equals sign in the regex should also be present in the regex for the dot separated variables (line 129)?

@jeff-schram Ah yeah, good point. Would you mind adding it?

**Comment by jeff-schram** [comment]
> > @pugnascotia Hey Rory, was working on that as well. I was also wondering if the trailing equals sign in the regex should also be present in the regex for the dot separated variables (line 129)?
> 
> @jeff-schram Ah yeah, good point. Would you mind adding it?

Nope, will add and push it within the hour or so

**Comment by pugnascotia** [comment]
@elasticmachine update branch

**Comment by jeff-schram** [comment]
> > > @pugnascotia Hey Rory, was working on that as well. I was also wondering if the trailing equals sign in the regex should also be present in the regex for the dot separated variables (line 129)?
> > 
> > 
> > @jeff-schram Ah yeah, good point. Would you mind adding it?
> 
> @pugnascotia Nope, no problem at all. Will add and push it within the hour or so

@pugnascotia I think last build answers the question, and the answer is no, I removed the trailing equals sign from the first regex.

**Comment by pugnascotia** [comment]
I'm not 100% sure those failures are related, I'm investigated right now.

File: distribution/src/bin/elasticsearch-env [hunk 1]
```diff
@@ -119,24 +119,25 @@ if [[ "$ES_DISTRIBUTION_TYPE" == "docker" ]]; then
...
-      fi
+  # Elasticsearch settings need to either:
+  # a. have at least two dot separated lower case words, e.g. `cluster.name`, or
+  while IFS='=' read -r envvar_key envvar_value; do
+    if [[ -n "$envvar_value" ]]; then
+      es_opt="-E${envvar_key}=${envvar_value}"
+      es_arg_array+=("${es_opt}")
     fi
-  done <<< "$(env)"
+  done <<< "$(env | grep -E '^[a-z0-9_]+\.[a-z0-9_]+')"
```
Comment (pugnascotia) [inline distribution/src/bin/elasticsearch-env:129]: Ah wait, the regex before just checked for (roughly) `^\w+\.\w+`, but actually there can be many periods in the name. Maybe we can mirror what we now have for the `ES_SETTING` variant e.g.

```suggestion
  done <<< "$(env | grep -E '^[a-z0-9_]+(\.[a-z0-9_]+)+=')"
```

**Comment by jeff-schram** [comment]
> I'm not 100% sure those failures are related, I'm investigated right now.

I think they are. I now have a test script locally that tests against the loop as in _master_ and against the changes in my branch. With my proposed change the ```xpack.security.enabled``` variable is missing from the final array whereas in _master_ it is not.  The hint for the first regex is in the comment above "have *at least*" two dot separated lower case words" and that is what the regex is testing for.

However, the following regex _does_ seem to work: ```([a-z0-9_]+\.[a-z0-9_]+)+=')``` (note the enclosing parentheses with ```+=``` appended, which is also done in the second regex).

**Comment by pugnascotia** [comment]
Haha, we're swapping. Yes, let's re-add the `=` but add grouping like you say.

File: distribution/src/bin/elasticsearch-env (hunk 1 above)
Comment (jeff-schram) [reply distribution/src/bin/elasticsearch-env:129]: Just tested your proposal locally (note that I am only testing the variables from the tests that failed thus far), and this works as well. What is the protocal? Can/should I click the _Commit suggestion_ button?

Comment (pugnascotia) [reply distribution/src/bin/elasticsearch-env:129]: Yes that works, or you can push the change yourself. It's up to you really, we'll do a squash commit anyway when we merge 👍

**Comment by jeff-schram** [comment]
Is the _elasticsearch-ci/part-2_ failure caused by our change? If so, any suggestion as how I can reproduce this locally for further debugging/development?

**Comment by pugnascotia** [comment]
@jeff-schram I don't think that's your change, no. I'm looking into it.

**Comment by pugnascotia** [comment]
@elasticmachine run elasticsearch-ci/part-2

**Comment by mark-vieira** [comment]
> Is the elasticsearch-ci/part-2 failure caused by our change? If so, any suggestion as how I can reproduce this locally for further debugging/development?

I suspect it might be. What we're doing here is bringing up Elasticsearch in a Docker container to use as a test fixture. Part of that is configuring a whole [host of settings via environment variables](https://github.com/elastic/elasticsearch/blob/308aee283dafd1765d942bd6470ef8a18d0340c6/x-pack/test/idp-fixture/docker-compose.yml#L5). I suspect we might be hitting an edge case where with some settings possible not being picked up properly.

**Comment by jeff-schram** [comment]
> I suspect it might be. What we're doing here is bringing up Elasticsearch in a Docker container to use as a test fixture. Part of that is configuring a whole [host of settings via environment variables](https://github.com/elastic/elasticsearch/blob/308aee283dafd1765d942bd6470ef8a18d0340c6/x-pack/test/idp-fixture/docker-compose.yml#L5). I suspect we might be hitting an edge case where with some settings possible not being picked up properly.
> 
> You'll need Docker and `docker-compose` installed locally, but you can reproduce like so:
> 
> ```shell
> ./gradlew :x-pack:qa:oidc-op-tests:check
> ```

It definitely was. I was able to reproduce it and found the issue: The backslash in the grouping is the culprit. _Any_ variable with multiple dots in its name were not passing through. Removing the backslash fixed the issue: the ```./gradlew :x-pack:qa:oidc-op-tests:check``` passes now without issues.

Luckily (considering the impact) we ran into this anyway, but maybe this would be a good addition to the ```DockerTests``` class?

On another note. On line 47 of the _docker-compose.yml_ there is a space in the variable value. Is that correct? My own test script choked on it and I had to take it out of the "test set", but it is suspect.

File: distribution/src/bin/elasticsearch-env [hunk 2]
```diff
@@ -126,7 +126,7 @@ if [[ "$ES_DISTRIBUTION_TYPE" == "docker" ]]; then
       es_opt="-E${envvar_key}=${envvar_value}"
       es_arg_array+=("${es_opt}")
     fi
-  done <<< "$(env | grep -E '^[a-z0-9_]+(\.[a-z0-9_]+)+=')"
+  done <<< "$(env | grep -E '^[a-z0-9_]+(.[a-z0-9_]+)+=')"
```
Comment (pugnascotia) [inline distribution/src/bin/elasticsearch-env:129]: Why does removing the backslash fix the issue with the oidc tests? Are we not simply matching any character with `.` now?

Comment (jeff-schram) [reply distribution/src/bin/elasticsearch-env]: Quite right. I was confused with the literal interpretation of dots within _square_ brackets. I undid that change. Instead, I had to update the regex to allow for _dashes_ in the variable name. That worked for my local test script and with the gradle task you provided.

Comment (pugnascotia) [reply distribution/src/bin/elasticsearch-env]: Nice, let's see what CI says then.

Comment (jeff-schram) [reply distribution/src/bin/elasticsearch-env]: I pushed the changes, but this change would make the regex quite a bit more strict. Before the variable would just have to _start_ with the character classes as specified and it didn't care what came after it. I am not sure if that was intended, but making the regex more strict would have impact on users with variable names containing any characters beyond the ```[-a-z0-9_]``` limitation as pushed in my last commit.

**Comment by mark-vieira** [comment]
> Luckily (considering the impact) we ran into this anyway, but maybe this would be a good addition to the DockerTests class?

Agreed. We should probably expand that scenario as implicitly relying on the fact that we use our Docker image in test fixtures isn't ideal.

**Comment by jeff-schram** [comment]
Copy/pasted this out of the resolved conversation to make sure it is not missed:

I pushed the changes, but this change will make the regex quite a bit more strict. Before the variable would just have to start with the character classes as specified and it didn't care what came after it. I am not sure if that was intended, but making the regex more strict will have impact on users with variable names containing any characters beyond the [-a-z0-9_] limitation as pushed in my last commit.

**Comment by mark-vieira** [comment]
> On another note. On line 47 of the docker-compose.yml there is a space in the variable value. Is that correct?

I'm not sure if it's "correct" but we should certainly support setting values with a space.

**Comment by jeff-schram** [comment]
> > On another note. On line 47 of the docker-compose.yml there is a space in the variable value. Is that correct?
> 
> I'm not sure if it's "correct" but we should certainly support setting values with a space.

Dove into that and [as long as certain characters are not present in the string it does not have to be quoted](https://symfony.com/doc/current/components/yaml/yaml_format.html#strings).

**Comment by jeff-schram** [comment]
Just wondering: how come the variables with dashes worked with the old regex as it was not part of the character classes, any how that came to be?

**Comment by pugnascotia** [comment]
I think it worked before because the pattern wasn't trying to match the whole string, and was actually just matching (roughly) `^\w+\.\w+.*` i.e. anything could follow if the first bit matched. Presumably we never tried set set something like `foo-bar.baz`, because it wouldn't have worked!

**Comment by pugnascotia** [comment]
Entertainingly, the `ES_SETTING_` mechanism doesn't allow hyphens, but I don't think I want to change that.

**Review by pugnascotia — APPROVED** [review]

**Comment by pugnascotia** [comment]
Thank you @jeff-schram!

Backported to `7.x` in 08ee0952c9e and `7.15` in 7a873191b2b.

**Pull request merged**